functional_programming_tests/test/main.cpp

```diff
@@ -0,0 +1,250 @@
+ public:
+  Map(){};
+  Map(std::vector<std::vector<T>> data) : data_{data} {};
+  T at(Position const& pos) const { return data_.at(pos.y).at(pos.x); }
+  std::vector<std::vector<T>>& get_data() { return data_; }
+  const std::vector<std::vector<T>>& get_data() const { return data_; }
+  bool empty() {return data_.empty();}
+
+ private:
+  std::vector<std::vector<T>> data_;
+};
+
+using Path = std::vector<Position>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, std::vector<Position> const& path) {
+  for (auto const& pose : path) {
+    os << "(" << pose.x << ", " << pose.y << ")\n";
+  }
+  return os;
+}
+
+// Operator overload for position comparison
+bool operator==(Position const& lhs, Position const& rhs) {
+  return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+namespace PathGenerator {
+
+  std::optional<Path> generate_global_path(
+      Position const& start, Position const& goal,
+      Map<unsigned char> const& costmap) {  // Calculation
+    // Some cool and nifty algorithm
+    // What is the delta in position
+    int const del_x = goal.x - start.x;
+    int const del_y = goal.y - start.y;
+
+    // What direction to move in for each dimension
+    int const del_x_sign = std::copysign(1.0, del_x);
+    int const del_y_sign = std::copysign(1.0, del_y);
+
+    // Push start onto the path
+    Path path;
+    path.push_back(start);
+
+    auto is_occupied = [&costmap](auto const x, auto const y) -> bool {
+      return costmap.at(Position{x, y}) == 255;
+    };
+
+    // Fails if there is any obstacle in the way
+    // Move horizontally
+    for (size_t i = 0; i < (std::abs(del_x)); ++i) {
+      if (is_occupied(path.back().x + del_x_sign, path.back().y)) {
+        return std::nullopt;
+      }
+      path.push_back({path.back().x + del_x_sign, path.back().y});
+    }
+    // Move vertically
+    for (size_t i = 0; i < (std::abs(del_y)); i++) {
+      if (is_occupied(path.back().x, path.back().y + del_y_sign)) {
+        return std::nullopt;
+      }
+      path.push_back({path.back().x, path.back().y + del_y_sign});
+    }
+
+    return path;
+  }  
+
+};
+
+namespace PathGeneratorUtilities {
+
+  std::optional<Map<unsigned char>> parseSetMapRequest(const std::shared_ptr<example_srvs::srv::SetMap::Request> request ) {
+    const auto costmap = request->map;
+    Map<unsigned char> map;
+    // Get the costmap from a ros topic
+    // Check that map layout makes sense
+    if ((costmap.layout.dim[0].size * costmap.layout.dim[1].size) !=
+        costmap.layout.dim[0].stride) {
+      return std::nullopt;
+    }
+    if (costmap.layout.dim[0].stride != costmap.data.size()) {
+      return std::nullopt;
+    }
+    auto const begin = std::begin(costmap.data);
+    // Populate the map
+    for (size_t row = 0; row < costmap.layout.dim[0].size; row++) {
+      map.get_data().push_back(
+          {begin + row * costmap.layout.dim[1].size,
+           begin + (row + 1) * costmap.layout.dim[1].size});
```

Comment:
Can we rename these things something like
```suggestion
          {row_beginning, row_end});
```
This seems like a perfect example of something that should be it's own function or operator that has it's own test. I'm a little skeptical of this put that's probably because I'm less familiar with the syntax

Reply:
Hmm, okay. I'll think about it.